### PR TITLE
xmlrpc-c: update to 1.54.06 and fix the abyss feature

### PR DESCRIPTION
**Pull request by yegorich**
Maintainer: @thess
Compile tested: ARMv7, TI Sitara AM335x, OpenWrt v23.05.2
Run tested: ARMv7, TI Sitara AM335x, OpenWrt v23.05.2, tested with internal software

Description:

Update version to 1.54.06 i.e. the latest version available as a tarball.

This PR also fixes the following build error breaking the abyss components:

xmlrpc_server_abyss.c:783:13: warning: format not a string literal and no format arguments [-Wformat-security]
  783 |             xmlrpc_faultf(envP, error);

### CODE REVIEW

File: libs/xmlrpc-c/Makefile [hunk 1]
```diff
@@ -8,12 +8,13 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=xmlrpc-c
-PKG_VERSION:=1.51.07
+PKG_VERSION:=1.60.02
```
Comment (BKPepe) [inline libs/xmlrpc-c/Makefile:11]: Why do we need to have advanced version based on https://xmlrpc-c.sourceforge.io/downloading.php ?

**Review by BKPepe — CHANGES_REQUESTED** [review]

File: libs/xmlrpc-c/Makefile (hunk 1 above)
Comment (yegorich) [reply libs/xmlrpc-c/Makefile:11]: We don't. I just wanted to keep the branch consistent with [Buildroot](https://github.com/buildroot/buildroot/blob/master/package/libxmlrpc/libxmlrpc.mk#L9). Should I switch to 1.59.02?

Comment (yegorich) [reply libs/xmlrpc-c/Makefile:11]: I was trying hard to avoid adding a patch :-) But I can do it.

**Comment by predators46** [comment]
@yegorich 

i tried to make it like this.

it looks like the libxmlrpc_xmltok file no longer exists in version 1.60.02

```
PKG_NAME:=xmlrpc-c
PKG_VERSION:=1.60.02
PKG_SOURCE_VERSION:=r3184
PKG_RELEASE:=1

PKG_SOURCE:=$(PKG_NAME)-code-$(PKG_SOURCE_VERSION)-advanced.zip
PKG_SOURCE_URL:=https://sourceforge.net/code-snapshots/svn/x/xm/xmlrpc-c/code/
PKG_HASH:=e3f8f3faa70102731f99419c4237b9a862bb65ca78100cd1c3b35039af1207df

PKG_MAINTAINER:=Ted Hess <thess@kitschensync.net>
PKG_LICENSE:=VARIOUS
PKG_LICENSE_FILES:=doc/COPYING

PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)-code-$(PKG_SOURCE_VERSION)-advanced
```



```
cp: cannot stat '/home/open/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced/ipkg-install/usr/lib/libxmlrpc_xmltok.so*': No such file or directory
make[2]: *** [Makefile:224: /home/open/armv8/bin/packages/aarch64_generic/packages/xmlrpc-c-internal_1.60.02-1_aarch64_generic.ipk] Error 1
make[2]: Leaving directory '/home/open/armv8/feeds/packages/libs/xmlrpc-c'
time: package/feeds/packages/xmlrpc-c/libxml2/compile#16.24#3.29#19.49
    ERROR: package/feeds/packages/xmlrpc-c failed to build (build variant: libxml2).
make[1]: *** [package/Makefile:128: package/feeds/packages/xmlrpc-c/compile] Error 1
make[1]: Leaving directory '/home/open/armv8'
make: *** [/home/open/armv8/include/toplevel.mk:225: package/xmlrpc-c/compile] Error 2
open@localhost:~/armv8$ cd /home/open/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced/
open@localhost:~/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced$ ls
aclocal.m4     doc                   missing             transport_config.mk
common.mk      dylib-common.mk       mkinstalldirs       unix-common.mk
config.guess   examples              README              version.h
config.log     GNUmakefile           shell_config        version.mk
config.mk      include               src                 Windows
config.mk.in   install-sh            srcdir.mk           xmlrpc-c-config
config.status  ipkg-aarch64_generic  srcdir.mk.in        xmlrpc-c-config.main
config.sub     ipkg-install          stamp-h.in          xmlrpc-c-config.test
configure      irix-common.mk        test                xmlrpc-c-config.test.main
configure.in   lib                   tools               xmlrpc_config.h
dll-common.mk  Makefile              transport_config.h  xmlrpc_config.h.in
open@localhost:~/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced$ cat version.mk
###############################################################################
# This is not only a make file inclusion, but a source file for the program
# 'mkvers' in the Windows build system.  The latter is very particular about
# the format of this file.  Do not change white space, add comment lines, or
# anything!
#
#
###############################################################################
XMLRPC_MAJOR_RELEASE = 1
XMLRPC_MINOR_RELEASE = 60
XMLRPC_POINT_RELEASE = 2
open@localhost:~/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced$ cd /home/open/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced/ipkg-install/usr/lib/
open@localhost:~/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced/ipkg-install/usr/lib$ ls
libxmlrpc.a               libxmlrpc_openssl.so            libxmlrpc_server.so.3.60
libxmlrpc_abyss.a         libxmlrpc_openssl.so.1          libxmlrpc.so
libxmlrpc_abyss.so        libxmlrpc_openssl.so.1.60       libxmlrpc.so.3
libxmlrpc_abyss.so.3      libxmlrpc_server.a              libxmlrpc.so.3.60
libxmlrpc_abyss.so.3.60   libxmlrpc_server_abyss.a        libxmlrpc_util.a
libxmlrpc_client.a        libxmlrpc_server_abyss.so       libxmlrpc_util.so
libxmlrpc_client.so       libxmlrpc_server_abyss.so.3     libxmlrpc_util.so.4
libxmlrpc_client.so.3     libxmlrpc_server_abyss.so.3.60  libxmlrpc_util.so.4.60
libxmlrpc_client.so.3.60  libxmlrpc_server.so             pkgconfig
libxmlrpc_openssl.a       libxmlrpc_server.so.3
open@localhost:~/armv8/build_dir/target-aarch64_generic_musl/xmlrpc-c-code-r3184-advanced/ipkg-install/usr/lib$
```

**Comment by jefferyto** [comment]
@predators46 the code snapshot link is not permanent; the code snapshot downloads are generated on-demand (from the user clicking the "Download Snapshot" link) and removed after a time. See https://sourceforge.net/p/forge/feature-requests/165/

**Comment by yegorich** [comment]
@predators46 this is not xmlrpc-c's fault. The recipe has to be changed to allow an XML backend selection, either internal or libxml2. Because now even if you select libxml2, the internal install section will be executed as well. 

I'll take care of it.

**Comment by yegorich** [comment]
I created a draft PR (https://github.com/openwrt/packages/pull/22783) solving the XML backend problem. The next step would be to update the package to the latest "archive" version. Aftet this one can try to solve the SVN update issue.

**Comment by yegorich** [comment]
@BKPepe @jefferyto I reworked this PR to update to 1.54.06.

**Comment by jefferyto** [comment]
> @BKPepe @jefferyto I reworked this PR to update to 1.54.06.

Looking good :+1: (spelling mistake in the file name "001-fix-format-paramter.patch" though)

Here's what I think you can do to get the libxml2 variant working:

* Remove the xmlrpc-c package (remove the `Package/xmlrpc-c` and `Package/xmlrpc-c/install` sections, the `BuildPackage` eval line)
* Uncomment `PROVIDES:=xmlrpc-c` in `Package/xmlrpc-c-internal` and `Package/xmlrpc-c-libxml2`
* Remove `@BROKEN` from `Package/xmlrpc-c-libxml2` and uncomment the `BuildPackage` eval line

I think you can also remove the `HIDDEN:=1` lines, I don't think they are necessary.

And add this patch (I suggest "002-remove-unnecessary-linking.patch"):

```patch
--- a/src/Makefile
+++ b/src/Makefile
@@ -183,7 +183,7 @@ $(LIBXMLRPC_SERVER): \
   $(call shliblefn, libxmlrpc)
 $(LIBXMLRPC_SERVER): LIBOBJECTS = $(LIBXMLRPC_SERVER_MODS:%=%.osh)
 $(LIBXMLRPC_SERVER): LIBDEP = \
-   -L. -lxmlrpc $(XML_PARSER_LIBDEP) $(LIBXMLRPC_UTIL_LIBDEP)
+   -L. -lxmlrpc $(LIBXMLRPC_UTIL_LIBDEP)
 
 LIBXMLRPC_SERVER_ABYSS = $(call shlibfn, libxmlrpc_server_abyss)
 
@@ -197,7 +197,7 @@ $(LIBXMLRPC_SERVER_ABYSS): LIBOBJECTS =
 $(LIBXMLRPC_SERVER_ABYSS): LIBDEP = \
   -L. -lxmlrpc_server \
   -L$(LIBXMLRPC_ABYSS_DIR) -lxmlrpc_abyss \
-  -L. -lxmlrpc $(XML_PARSER_LIBDEP) $(LIBXMLRPC_UTIL_LIBDEP)
+  -L. -lxmlrpc $(LIBXMLRPC_UTIL_LIBDEP)
 ifeq ($(MSVCRT),yes)
   $(LIBXMLRPC_SERVER_ABYSS):  LIBDEP += -lws2_32 -lwsock32
 endif
@@ -212,7 +212,7 @@ $(LIBXMLRPC_SERVER_CGI): \
 $(LIBXMLRPC_SERVER_CGI): LIBOBJECTS = $(LIBXMLRPC_SERVER_CGI_MODS:%=%.osh)
 $(LIBXMLRPC_SERVER_CGI): LIBDEP = \
   -L. -lxmlrpc_server \
-  -L. -lxmlrpc $(XML_PARSER_LIBDEP) $(LIBXMLRPC_UTIL_LIBDEP)
+  -L. -lxmlrpc $(LIBXMLRPC_UTIL_LIBDEP)
 
 LIBXMLRPC_CLIENT = $(call shlibfn, libxmlrpc_client)
 
@@ -237,7 +237,6 @@ $(LIBXMLRPC_CLIENT): LIBOBJECTS = \
 LIBXMLRPC_CLIENT_LIBDEP = \
   -Lblddir/src -Lblddir/lib/libutil \
   -lxmlrpc -lxmlrpc_util \
-  $(XML_PARSER_LIBDEP) \
   $(TRANSPORT_LIBDEP) \
 
 $(LIBXMLRPC_CLIENT): LIBDEP = \
```

I did a quick test build and it worked but I haven't done any run-testing. It would be nice to get this working though I don't think this is strictly necessary for this PR.

**Comment by yegorich** [comment]
@jefferyto I have applied your changes and now, I can compile either internal or libxml2 variant of all XMLRPC-C packages. I have also tested my software with both variants and it is also working.

**Comment by neheb** [comment]
@yegorich rtorrent resolves to the internal? variant and fails because of missing dependency. Can you look?

https://downloads.openwrt.org/snapshots/faillogs/i386_pentium-mmx/packages/rtorrent/rpc/compile.txt

**Comment by yegorich** [comment]
@neheb I cannot reproduce this issue. Could you provide a minimal .config that triggers this build failure?

**Comment by neheb** [comment]
Nope. Just select rtorrent-rpc as M

**Comment by yegorich** [comment]
@neheb I could reproduce the error. To do so, I had to select both `xmlrpc-c-internal` and `xmlrpc-c-libxml2`. I'll look deeper into the matter.

**Comment by jefferyto** [comment]
I haven't tested this, but my guess is because Build/InstallDev is run for both internal and libxml2 variants, perhaps the internal variant was run last and rtorrent-rpc linked against it.

**Comment by yegorich** [comment]
I wonder whether switching rtorrent to pkgconf would fix the problem. I prepared a PR for xmlrpc-c (https://github.com/openwrt/packages/pull/22995). My own software has no problem with both internal and external variants selected.

I have to rewrite rtorrent to use pkgconf for xmlrpc-c though.

**Comment by neheb** [comment]
That would be ideal.

rtorrent seems to be somewhat dead upstream.

**Comment by yegorich** [comment]
pkg-conf has nothing to do with this isssue. I could find the cause why my packages made no problems:

`DEPENDS:=+xmlrpc-c +xmlrpc-c-abyss +xmlrpc-c-server-abyss +pkg-config +curl`

I have `+xmlrpc-c` as a dependency. This selects `PACKAGE_xmlrpc-c-libxml2`. Don't know why though.

With this workaround I can also compile `rtorrent-rpc`.

**Pull request merged**